source/language/scope.rst

```diff
@@ -0,0 +1,338 @@
+Scoping of variables
+====================
+
+This section describes the rules surrounding scoping of variables in OpenQASM.
+
+OpenQASM 3.0 has four types of scoping constructs, and the visibility of some
+symbols can differ slightly between these.  The types of scope are:
+
+* :ref:`Global scope <scope-global>`, which is the current scope only when no
+  other scopes are active.
+
+* :ref:`Gate and function scope <scope-subroutine>`, which is entered in the
+  body of ``gate`` and ``def`` definitions.
+
+* :ref:`Local block scope <scope-block>`, which becomes active on entry to a
+  non-gate and non-subroutine control-flow block, such as ``if`` or ``for``.
+
+* Calibration scopes, which are the shared calibration state contained in
+  ``cal`` and ``defcal`` blocks.  The precise rules for these may vary depending
+  on the particular companion calibration grammar that has been loaded for this
+  program.  See :ref:`the OpenPulse specification <openpulse>` for details on
+  how this is handled in the OpenPulse calibration language.  This document will
+  not cover this topic further.
+
+In general, the lifetime of each identifier begins when it is declared, and ends
+at the completion of the scope it was declared in.  The storage space for each
+variable must be allocated for at least the lifetime of the identifier, but it
+is up to individual implementations to define their allocation strategies within
+nested scopes.
+
+In order to be a valid OpenQASM 3.0 program, symbols must be defined before they
+are used; in particular, there is no forward definition of functions and gates,
+and there can be no mutual recursion.
+
+Most regular identifiers of variables can be shadowed in inner scopes, so that
+the identifier temporarily refers to a different variable, but cannot be
+re-declared without entering an inner scope.  The type of the shadowing variable
+does not need to be the same as the type of the variable it is shadowing.
+
+Not all identifiers declared in outer scopes are visible within inner scopes.
+The visibility depends on the type of the variable, and the type of the inner
+scope.  Approximately, ``const`` variables, gates and subroutines are visible
+within *all* inner scopes, while other variables in outer scopes are visible
+within inner control-flow scopes but not gate and function scopes.
+
+The ``include`` statement should be seen as extending the scope it is contained
+within; all variables that are in scope at the time of the ``include`` statement
```

Comment:
Make this clearer that the `include` is only global scope (and update the `include` section too).

Reply:
Done in 01ed7cc.